webhook/pr_review.go

```diff
@@ -0,0 +1,374 @@
+	}(r.Body)
+	if err != nil {
+		http.Error(w, "Error reading request body", http.StatusInternalServerError)
+		return
+	}
+
+	var event PullRequestEvent
+	err = json.Unmarshal(body, &event)
+	if err != nil {
+		http.Error(w, "Error unmarshaling request body", http.StatusInternalServerError)
+		return
+	}
+
+	if event.Action == "opened" || event.Action == "reopened" || event.Action == "synchronize" {
+		log.Println("Pull request event received:", event.Action)
+		owner, repo := parseRepoInfo(event.Repository.FullName)
+		var files []File
+		files, err = fetchPRFiles(owner, repo, event.Number)
+		if err != nil {
+			log.Println("Error fetching PR files:", err)
+			return
+		}
+
+		commitID, err := fetchPRCommitID(owner, repo, event.Number)
+		log.Println("Commit ID:", commitID)
+		if err != nil {
+			log.Println("Error fetching PR commit ID:", err)
+			return
+		}
+		submitReview(owner, repo, event.Number, commitID, files)
+	} else {
+		log.Println("Ignored event:", event.Action)
+	}
+}
+
+func initGithubParameter() {
+	apiBaseurl = viper.GetString("git.api_base_url")
+	token = viper.GetString("git.token")
+	owner = viper.GetString("git.owner")
+	repo = viper.GetString("git.repo")
+	log.Printf("apiBaseurl=%s, token=%s, owner=%s, repo=%s\n", apiBaseurl, token, owner, repo)
+	initFlag = true
+}
+
+func parseRepoInfo(fullName string) (owner, repo string) {
+	s := strings.Split(fullName, "/")
+	return s[0], s[1]
+}
+
+func fetchPRFiles(owner, repo string, prNumber int) ([]File, error) {
+	client := &http.Client{}
+
+	req, err := http.NewRequest("GET", fmt.Sprintf(fetchPRFilesURL, apiBaseurl, owner, repo, prNumber), nil)
+	if err != nil {
+		log.Println("Error creating request:", err)
+		return nil, err
+	}
+
+	req.Header.Set("Accept", "application/vnd.github+json")
+	req.Header.Set("Authorization", "token "+token)
+
+	resp, err := client.Do(req)
+	if err != nil {
+		log.Println("Error sending request:", err)
+		return nil, err
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode == http.StatusOK {
+		bodyBytes, err := io.ReadAll(resp.Body)
+		if err != nil {
+			log.Println("Error reading response body:", err)
+			return nil, err
+		}
+
+		var files []File
+		err = json.Unmarshal(bodyBytes, &files)
+		if err != nil {
+			log.Println("Error unmarshaling response body:", err)
+			return nil, err
+		}
+		return files, nil
+	} else {
+		log.Printf("Error fetching PR files: %d\n", resp.StatusCode)
+		return nil, fmt.Errorf("Error fetching PR files: %d", resp.StatusCode)
+	}
+}
+
+func fetchPRCommitID(owner, repo string, prNumber int) (string, error) {
+	client := &http.Client{}
+
+	req, err := http.NewRequest("GET", fmt.Sprintf(fetchCommitsURL, apiBaseurl, owner, repo, prNumber), nil)
+	if err != nil {
+		return "", err
+	}
+
+	req.Header.Set("Content-Type", "application/json")
+	req.Header.Set("Authorization", "token "+token)
+
+	resp, err := client.Do(req)
+	if err != nil {
+		return "", err
+	}
+	defer func(Body io.ReadCloser) {
+		err := Body.Close()
```

Comment:
1. The code does not have proper error handling for some of the HTTP requests, which can lead to unexpected behavior or crashes.
2. The code does not have input validation, which can lead to security vulnerabilities.
3. The code has hardcoded values for the GitHub API URLs and maximum code character number, which can be improved by using environment variables or configuration files instead.
4. The code can be improved by breaking down the functions into smaller, more modular functions with specific purposes.